fendermint/vm/interpreter/src/fvm/state/ipc.rs

```diff
@@ -189,54 +186,65 @@ impl<DB: Blockstore> GatewayCaller<DB> {
+    pub fn commit_parent_finality(
         &self,
-        finality: fendermint_vm_topdown::IPCParentFinality,
-    ) -> anyhow::Result<FvmMessage> {
+        state: &mut FvmExecState<DB>,
+        finality: IPCParentFinality,
+    ) -> anyhow::Result<Option<IPCParentFinality>> {
         let evm_finality = router::ParentFinality::try_from(finality)?;
-        let call = self.router.contract().commit_parent_finality(evm_finality);
-        let calldata = call
-            .calldata()
-            .ok_or_else(|| anyhow!("no calldata for commit parent finality"))?;
-
-        encode_to_fvm_implicit(calldata.as_ref())
+        let (has_committed, prev_finality) = self
+            .router
+            .call(state, |c| c.commit_parent_finality(evm_finality))?;
+        Ok(if !has_committed {
+            None
+        } else {
+            Some(IPCParentFinality::try_from(prev_finality)?)
+        })
     }
 
-    pub fn store_validator_changes_msg(
+    pub fn store_validator_changes(
         &self,
+        state: &mut FvmExecState<DB>,
         changes: Vec<StakingChangeRequest>,
-    ) -> anyhow::Result<FvmMessage> {
+    ) -> anyhow::Result<()> {
         let mut change_requests = vec![];
         for c in changes {
             change_requests.push(router::StakingChangeRequest::try_from(c)?);
         }
 
-        let call = self
-            .router
-            .contract()
-            .store_validator_changes(change_requests);
-        let calldata = call
-            .calldata()
-            .ok_or_else(|| anyhow!("no calldata for store validator changes"))?;
+        self.router
+            .call(state, |c| c.store_validator_changes(change_requests))
+    }
 
-        encode_to_fvm_implicit(calldata.as_ref())
+    /// Call this function to mint some FIL to the gateway contract
+    pub fn mint_to_gateway(
+        &self,
+        state: &mut FvmExecState<DB>,
+        value: TokenAmount,
+    ) -> anyhow::Result<()> {
+        let state_tree = state.state_tree_mut();
+        state_tree.mutate_actor(ipc::GATEWAY_ACTOR_ID, |actor_state| {
+            actor_state.balance += value;
+            Ok(())
+        })?;
+        Ok(())
     }
 
-    pub fn apply_cross_messages_msg(
+    pub fn apply_cross_messages(
         &self,
+        state: &mut FvmExecState<DB>,
         cross_messages: Vec<CrossMsg>,
-    ) -> anyhow::Result<FvmMessage> {
-        let mut messages = vec![];
-        for c in cross_messages {
-            messages.push(router::CrossMsg::try_from(c)?);
-        }
-
-        let call = self.router.contract().apply_cross_messages(messages);
-        let calldata = call
-            .calldata()
-            .ok_or_else(|| anyhow!("no calldata for apply cross messages"))?;
-
-        encode_to_fvm_implicit(calldata.as_ref())
+    ) -> anyhow::Result<FvmApplyRet> {
+        let messages = cross_messages
+            .into_iter()
+            .map(router::CrossMsg::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
```suggestion
            .collect::<Result<Vec<_>, _>>()
            .context("failed to convert cross messages")?;
```

I hope this cannot be triggered by malicious users.